connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java

```diff
@@ -760,12 +742,35 @@ void assignedOffsets(Map<TopicPartition, OffsetAndMetadata> offsets) {
             consumedOffsets = new HashMap<>(offsets);
             processedOffsets = offsets;
             sinkRecordLag.record(0.0);
+            // Add topic partition groups for each new topic partition and remove those not used anymore
+            Set<TopicPartition> keysToRemove = new HashSet<>(offsets.keySet());
+            for (TopicPartition topicPartition : offsets.keySet()) {
+                TopicPartitionMetricsGroup group = topicPartitionMetrics.get(topicPartition);
+                if (group == null) {
+                    group = new TopicPartitionMetricsGroup(id, metrics, topicPartition);
+                    topicPartitionMetrics.put(topicPartition, group);
+                }
+                group.recordLag(0.0);
+                keysToRemove.remove(topicPartition);
```

Comment:
something seems off here -- `keysToRemove` is initialized to `offsets.keySet()` but it looks like we'll just remove all of the same keys via this loop?

Reply:
When I rebase on the latest of #3975, this logic will go away.